crates/subspace-block-relay/src/consensus.rs

```diff
@@ -0,0 +1,622 @@
+use lru::LruCache;
+use parking_lot::Mutex;
+use sc_client_api::{BlockBackend, HeaderBackend};
+use sc_network::request_responses::{IncomingRequest, OutgoingResponse, ProtocolConfig};
+use sc_network::types::ProtocolName;
+use sc_network::{PeerId, RequestFailure};
+use sc_network_common::sync::message::{BlockAttributes, BlockData, BlockRequest, FromBlock};
+use sc_network_sync::block_relay_protocol::{
+    BlockDownloader, BlockRelayParams, BlockResponseError, BlockServer,
+};
+use sc_service::SpawnTaskHandle;
+use sc_transaction_pool_api::{InPoolTransaction, TransactionPool, TxHash};
+use sp_runtime::generic::BlockId;
+use sp_runtime::traits::{Block as BlockT, Header};
+use sp_runtime::Justifications;
+use std::num::NonZeroUsize;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tracing::{info, trace, warn};
+
+type BlockHash<Block> = <Block as BlockT>::Hash;
+type BlockHeader<Block> = <Block as BlockT>::Header;
+type Extrinsic<Block> = <Block as BlockT>::Extrinsic;
+
+const SYNC_PROTOCOL: &str = "/subspace/consensus-block-relay/1";
+
+// TODO: size these properly, or move to config
+const NUM_PEER_HINT: NonZeroUsize = NonZeroUsize::new(100).expect("Not zero; qed");
+const TRANSACTION_CACHE_SIZE: NonZeroUsize = NonZeroUsize::new(512).expect("Not zero; qed");
+
+/// Initial request to the server
+/// We currently ignore the direction field and return a single block,
+/// revisit if needed
+#[derive(Encode, Decode)]
+struct InitialRequest<Block: BlockT, ProtocolReq> {
+    /// Starting block
+    from_block: BlockId<Block>,
+
+    /// Requested block components
+    block_attributes: u32,
+
+    /// The protocol specific part of the request
+    protocol_request: ProtocolReq,
+}
+
+/// Initial response from server
+#[derive(Encode, Decode)]
+struct InitialResponse<Block: BlockT, ProtocolRsp> {
+    ///  Hash of the block being downloaded
+    block_hash: BlockHash<Block>,
+
+    /// The partial block, without the extrinsics
+    partial_block: PartialBlock<Block>,
+
+    /// The opaque protocol specific part of the response.
+    /// This is optional because BlockAttributes::BODY may not be set in
+    /// the BlockRequest, in which case we don't need to fetch the
+    /// extrinsics
+    protocol_response: Option<ProtocolRsp>,
+}
+
+/// The partial block response from the server. It has all the fields
+/// except the extrinsics. The extrinsics are handled by the protocol
+#[derive(Encode, Decode)]
+struct PartialBlock<Block: BlockT> {
+    hash: BlockHash<Block>,
+    header: Option<BlockHeader<Block>>,
+    indexed_body: Option<Vec<Vec<u8>>>,
+    justifications: Option<Justifications>,
+}
+
+/// The message to the server
+#[derive(Encode, Decode)]
+enum ServerMessage<Block: BlockT, ProtocolReq> {
+    /// Initial message, to be handled both by the client
+    /// and the protocol
+    InitialRequest(InitialRequest<Block, ProtocolReq>),
+
+    /// Message to be handled by the protocol
+    ProtocolReq(ProtocolReq),
+}
```

Comment:
Now I'm wondering why abstract protocol needs to distinguish between intial and other requests? It could be just protocol request, if specific protocol has different kinds of messages (initial, continuation, final) it can have an enum and compose them as desired.

Reply:
I know that implementation-wise there are several phases, what I don't understand is why do we need to expose them on abstract level. What I expect is that abstraction just calls into specific implementation to retrieve the block. How many requests it does and what those requests do does not matter, that matters is that we get `Result<Block, Error>` as the result. This looks like potentially leaking abstractions.